### PR TITLE
New User Welcome Email

### LINKED ISSUE #11: Email Validation for User Registration and Update
To ensure data security and ensure the validity of user emails, we need to implement an email validation system during user registration and email updates. This feature will involve sending a verification code to the user's email, requiring them to validate their email address.
Feature Ticket: Email Validation for User Registration and Update

**Requirements:**
- [x] The system should send an email containing a verification code to the user's provided email address.
- [x] A unique validation code should be generated for each email verification request.
- [x] The validation code should have a time limit of 5 minutes for its validity. If a new code is generated, the previous one should expire automatically.
- [x] Users who have not completed the email validation process should not be allowed to log in unless they are already logged in (existing sessions should not be interrupted).
- [x] In development and test environments, the email validation step should be skippable to facilitate testing and development processes.

### PR DESCRIPTION
**Pull request by chrisdedman**
Issue -> https://github.com/Code-Society-Lab/cursif-backend/issues/11

### CODE REVIEW

File: lib/cursif_web/plugs/rate_limit.ex
Comment (chrisdedman) [inline lib/cursif_web/plugs/rate_limit.ex:1]: I just remove the doc for the private function to remove the warning during compilation

File: config/runtime.exs [hunk 1]
```diff
@@ -68,16 +68,20 @@ if config_env() == :prod do
...
   # Also, you may need to configure the Swoosh API client of your choice if you
   # are not using SMTP. Here is an example of the configuration:
-  #
-  #     config :cursif, Cursif.Mailer,
-  #       adapter: Swoosh.Adapters.Mailgun,
-  #       api_key: System.get_env("MAILGUN_API_KEY"),
-  #       domain: System.get_env("MAILGUN_DOMAIN")
-  #
+  
+# config :cursif, Cursif.Mailer,
```
Comment (PenguinBoi12) [inline config/runtime.exs]: Why everything is commented here?

File: config/runtime.exs [hunk 2]
```diff
@@ -68,16 +68,20 @@ if config_env() == :prod do
...
+#   ssl: System.get_env("SMTP_SSL") || false
+#   tls: :always,
+#   port: System.get_env("SMTP_PORT")
+  
   # For this example you need include a HTTP client required by Swoosh API client.
   # Swoosh supports Hackney and Finch out of the box:
-  #
-  #     config :swoosh, :api_client, Swoosh.ApiClient.Hackney
+  
+  # config :swoosh, :api_client, Swoosh.ApiClient.Gmail
```
Comment (PenguinBoi12) [inline config/runtime.exs]: Originally it was ` Swoosh.ApiClient.Hackney`, can you please put it back.

File: config/config.exs
Comment (PenguinBoi12) [inline config/config.exs:1]: You should be able to reset this file from origin by running `git checkout origin/main lib/config/config.exs`

File: lib/cursif/accounts.ex [hunk 3]
```diff
@@ -119,25 +120,83 @@ defmodule Cursif.Accounts do
...
+  @spec authenticate_user(String.t(), String.t()) :: {:ok, User.t(), String.t()} | {:error, :invalid_credentials | :not_confirmed}
   def authenticate_user(email, plain_text_password) do
     case Repo.get_by(User, email: email) do
       nil ->
-        Argon2.no_user_verify()
         {:error, :invalid_credentials}
       user ->
-        if Argon2.verify_pass(plain_text_password, user.hashed_password) do
-          {:ok, user, create_token(user)}
+        if user.confirmed_at == nil do
```
Comment (PenguinBoi12) [inline lib/cursif/accounts.ex]: You can use `is_nil(user.confirmed_at)`

File: lib/cursif/accounts.ex [hunk 4]
```diff
@@ -119,25 +120,83 @@ defmodule Cursif.Accounts do
...
     case Cursif.Guardian.encode_and_sign(user, %{}) do
       {:ok, token, _full_claims} -> token
     end
   end
+
+  def create_token(user, :phoenix_token) do
+    {:ok, Phoenix.Token.sign(CursifWeb.Endpoint, "user id", user.id)}
+  end
+
+  def generate_validation_token(user) do
```
Comment (PenguinBoi12) [inline lib/cursif/accounts.ex]: The point of `create_token(user, :phoenix_token)` is not to have to do this

File: lib/cursif_web.ex [hunk 5]
```diff
@@ -37,6 +37,12 @@ defmodule CursifWeb do
       import Phoenix.Controller,
         only: [get_flash: 1, get_flash: 2, view_module: 1, view_template: 1]
 
+      # Use all HTML functionality (forms, tags, etc)
```
Comment (PenguinBoi12) [inline lib/cursif_web.ex]: ` import CursifWeb.ErrorHelpers` and ` import CursifWeb.Gettext` don't need to be imported and `should be imported in `view_helper`

File: lib/cursif/accounts.ex [hunk 6]
```diff
@@ -119,25 +120,83 @@ defmodule Cursif.Accounts do
...
+
+  ## Examples
+
+      iex> generate_confirmation_token(user)
+      {:ok, %User{}}
+
+      iex> generate_confirmation_token(user)
+      {:error, :already_confirmed}
+  """
+  def generate_confirmation_token(%User{} = user) do
```
Comment (PenguinBoi12) [inline lib/cursif/accounts.ex]: I don't like that it's called `generate_confirmation_token` but also sends confirmation. Maybe `send_confirmation_email` would be better.

File: lib/cursif_web/templates/email/welcome.html.eex [hunk 7]
```diff
@@ -0,0 +1,85 @@
+<!DOCTYPE html>
+<html>
+  <link rel="stylesheet" type="text/css" href="//fonts.googleapis.com/css?family=Montez" />
+  <link rel="stylesheet" type="text/css" href="//fonts.googleapis.com/css?family=Roboto" />
+  <style>
+    body {
```
Comment (PenguinBoi12) [inline lib/cursif_web/templates/email/welcome.html.eex:6]: Have you looked/tried to put that in the `assets/css`?

File: test/support/fixtures/accounts_fixtures.ex [hunk 8]
```diff
@@ -47,7 +48,9 @@ defmodule Cursif.AccountsFixtures do
   """
   def create_unique_user(_ \\ %{}) do
     user_attrs = unique_user_attributes()
+
     {:ok, user} = Accounts.create_user(user_attrs)
+    {:ok, user} = User.confirm_user_email(user)
```
Comment (PenguinBoi12) [inline test/support/fixtures/accounts_fixtures.ex]: `User.confirm_user_email` is slightly redundant. Maybe `User.confirm_email` is enough?

**Review by PenguinBoi12 — CHANGES_REQUESTED** [review]
There's one big issue. When we validate, we're redirected to the backend, that's not really good, we should return the user to the client. In your issue you don't have to implement the client part but you have to think about it.

In a rest API we might want a `api/users/confirm/:token/` but in our case, since our API is a graphql API, it would most likely be cleaner to created a resolver that will verify the email. So, you most likely don't need any controller (that's my bad, I should've clicked when you talked about controllers)

To get the client's url we might be able to fetch the origin from the request. Otherwise, we could have an optional parameter passed containing the client's url or create client token (yes more tokens hahah).

So basically, instead of building the verify with the backend url it needs to be done with the client's url. The rest is good (except the route might be simplified by just calling it `verify` or something like that)


There's one thing that has been forgotten. The seeds, they create user but none of them are validated haha. It would probably be nice to validate them as we create them so we can use them. It makes me think... what if we lose the email? How can we trigger another email? I think another resolver could be added for that. 

In term of code, you did some good job. Well done! Ofc, some thing will change with the other token system but you don't have to worry about that.

File: config/config.exs
Comment (chrisdedman) [reply config/config.exs:1]: This file has been restored from previous edition, with my new configuration as well.

File: lib/cursif_web/templates/email/welcome.html.eex (hunk 7 above)
Comment (chrisdedman) [reply lib/cursif_web/templates/email/welcome.html.eex:6]: I am going to try that

File: lib/cursif_web.ex (hunk 5 above)
Comment (chrisdedman) [reply lib/cursif_web.ex:40]: Yes, you are right. Fixed

File: test/support/fixtures/accounts_fixtures.ex (hunk 8 above)
Comment (chrisdedman) [reply test/support/fixtures/accounts_fixtures.ex]: Yes, you are right. Fixed!

File: config/runtime.exs (hunk 1 above)
Comment (chrisdedman) [reply config/runtime.exs]: If I uncomment them, I cannot run the project locally.

File: config/config.exs [hunk 9]
```diff
@@ -63,11 +65,15 @@ config :phoenix, :json_library, Jason
...
+  secret_key: Application.get_env(:cursif, :secret_key)
 
 config :cursif, env: Mix.env()
 # Import environment specific config. This must remain at the bottom
 # of this file so it overrides the configuration defined above.
 import_config "#{config_env()}.exs"
+
+# Get the client URL from the environment or use a default
+config :cursif, 
+  client_url: System.get_env("CLIENT_URL") || "http://localhost:3000"
```
Comment (PenguinBoi12) [inline config/config.exs]: It's better to do ` System.get_env("CLIENT_URL", "http://localhost:3000")`. Also, `config :cursif` is already defined at the top. Define it there instead. 

However, in my opinion it should be placed in `dev.exs`, that way we have a default only on dev (which is what we want; we want to force prod to define it)

File: config/runtime.exs [hunk 10]
```diff
@@ -68,16 +68,24 @@ if config_env() == :prod do
...
   # Swoosh supports Hackney and Finch out of the box:
-  #
-  #     config :swoosh, :api_client, Swoosh.ApiClient.Hackney
+  
+  config :swoosh, :api_client, Swoosh.ApiClient.Hackney
   #
   # See https://hexdocs.pm/swoosh/Swoosh.html#module-installation for details.
+
+  # Get the client URL from the environment or use a default
+  config :cursif,
```
Comment (PenguinBoi12) [inline config/runtime.exs]: Technically, it's already defined in `config.exs` but I've already talked about that in another comment. But this should all be in `prod.exs`. You shouldn't really modify `runtime.exs`

However, it's fine that it's defined isn't that bad although it should raise an error if undefined.

File: config/prod.exs [hunk 11]
```diff
@@ -49,4 +49,8 @@ config :logger, level: :info
 # Check `Plug.SSL` for all available options in `force_ssl`.
 
 # execute `mix guardian.gen.secret` and that you set it in your environment variables under `CURSIF_SECRET_KEY`.
-config :cursif, :secret_key, System.fetch_env!("CURSIF_SECRET_KEY")
\ No newline at end of file
+config :cursif, :secret_key, System.fetch_env!("CURSIF_SECRET_KEY")
+
+# Get the client URL from the environment or use a default
+config :cursif, 
+  client_url: System.get_env("CLIENT_URL_PROD") || "http://localhost:3000"
```
Comment (PenguinBoi12) [inline config/prod.exs]: `CLIENT_URL_PROD` the key should be the same as the other.   There's not use case of needing both at the same time.

File: config/dev.exs [hunk 12]
```diff
@@ -65,3 +65,11 @@ config :phoenix, :plug_init_mode, :runtime
...
 # See Guardian setup in `config/config.exs`
 config :cursif, :secret_key, "dev_secret_key"
+
+# Configure the mailer
+config :cursif, Cursif.Mailer,
+  adapter: Swoosh.Adapters.Local
+
+# Get the client URL from the environment or use a default
+config :cursif, 
+  client_url: System.get_env("CLIENT_URL_DEV") || "http://localhost:3000"
```
Comment (PenguinBoi12) [inline config/dev.exs]: `CLIENT_URL_PROD` the key should be the same as the other.   There's not use case of needing both at the same time.

File: lib/cursif_web/emails/user_email.ex [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    view: CursifWeb.EmailView, 
+    layout: {CursifWeb.LayoutView, :email}
+
+  import Swoosh.Email
+  alias Cursif.Mailer
+
+  defp send_email(user, subject, base_url) do
+    email =
+      new()
+      |> from({"Cursif", "cursif@codesociety.xyz"})
```
Comment (PenguinBoi12) [inline lib/cursif_web/emails/user_email.ex]: When I think about it... This might also need to be a configuration.

File: lib/cursif_web/plugs/rate_limit.ex [hunk 14]
```diff
@@ -14,9 +14,6 @@ defmodule CursifWeb.RateLimit do
     rate_limit(conn)
   end
 
-  @doc """
```
Comment (PenguinBoi12) [inline lib/cursif_web/plugs/rate_limit.ex:17]: Why did you remove that comment?

File: lib/cursif_web/views/layout_view.ex
Comment (PenguinBoi12) [inline lib/cursif_web/views/layout_view.ex:1]: Is it used or needed?

**Review by PenguinBoi12 — CHANGES_REQUESTED** [review]
In general it's pretty good, the only issue is the configs. They are a bit of a mess. Otherwise, after that it should be fine. Well done!

File: lib/cursif_web/plugs/rate_limit.ex (hunk 14 above)
Comment (chrisdedman) [reply lib/cursif_web/plugs/rate_limit.ex:17]: because it generate a warning at compiler time (because it is a private function)

File: lib/cursif_web/views/layout_view.ex
Comment (chrisdedman) [reply lib/cursif_web/views/layout_view.ex:1]: It is needed. Without it, it does not work

File: config/config.exs [hunk 15]
```diff
@@ -9,7 +9,9 @@ import Config
 
 config :cursif,
   ecto_repos: [Cursif.Repo],
-  generators: [binary_id: true]
+  generators: [binary_id: true],
+  client_url: System.get_env("CLIENT_URL"),
+  email_address: "cursif@codesociety.xyz"
```
Comment (PenguinBoi12) [inline config/config.exs]: I meant this should be a env var too.

**Review by PenguinBoi12 — APPROVED** [review]
